stream_alert/threat_intel_downloader/threat_stream.py

```diff
@@ -0,0 +1,273 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from datetime import datetime, timedelta
+import json
+
+import backoff
+import boto3
+from botocore.exceptions import ClientError
+import requests
+
+from stream_alert.shared.backoff_handlers import (
+    backoff_handler,
+    success_handler,
+    giveup_handler
+)
+
+from stream_alert.threat_intel_downloader import LOGGER
+from stream_alert.threat_intel_downloader.exceptions import (
+    ThreatStreamCredsError,
+    ThreatStreamRequestsError
+)
+
+class ThreatStream(object):
+    """Class to retrieve IOCs from ThreatStream.com and store them in DynamoDB"""
+    _API_URL = 'https://api.threatstream.com'
+    _API_RESOURCE = 'intelligence'
+    _IOC_STATUS = 'active'
+    # max IOC objects received from one API call, default is 0 (equal to 1000)
+    _API_MAX_LIMIT = 1000
+    _API_MAX_INDEX = 1000000
+    _PARAMETER_NAME = 'threat_intel_downloader_api_creds'
+
+    def __init__(self, config):
+        self.ioc_types = config['ioc_types']
+        self.ioc_sources = config['ioc_filters']
+        self.threshold = self._API_MAX_INDEX - self._API_MAX_LIMIT
+        self.region = config['region']
+        self.ioc_keys = config['ioc_keys']
+        self.api_user = None
+        self.api_key = None
+        self._get_api_creds()
+        self.table_name = config['function_name']
+
+    def _get_api_creds(self):
+        """Retrieve ThreatStream API credentials from Parameter Store"""
+        try:
+            ssm = boto3.client('ssm', self.region)
+            response = ssm.get_parameters(
+                Names=[self._PARAMETER_NAME], WithDecryption=True
+            )
+        except ClientError as err:
+            LOGGER.error('SSM client error: %s', err)
+            raise
+
+        for cred in response['Parameters']:
+            if cred['Name'] == self._PARAMETER_NAME:
+                try:
+                    decoded_creds = json.loads(cred['Value'])
+                    self.api_user = decoded_creds['api_user']
+                    self.api_key = decoded_creds['api_key']
+                except ValueError:
+                    LOGGER.error('Can not load value for parameter with '
+                                 'name \'%s\'. The value is not valid json: '
+                                 '\'%s\'', cred['Name'], cred['Value'])
+                    raise ThreatStreamCredsError('ValueError')
+
+        if not (self.api_user and self.api_key):
+            LOGGER.error('API Creds Error')
+            raise ThreatStreamCredsError('API Creds Error')
+
+    exceptions_to_backoff = (requests.exceptions.Timeout,
+                             requests.exceptions.ConnectionError,
+                             requests.exceptions.ChunkedEncodingError,
+                             ThreatStreamRequestsError)
+    @backoff.on_exception(backoff.constant,
+                          exceptions_to_backoff,
```

Comment:
I thought about using `expo`, but I don't want to add too much timeout. I can change to `expo` if constant doesn't work well. Using constant with 3 times retries, the max timeout will be 3 seconds. If use `expo`, it will take `1+2+4 = 7` seconds.

Reply:
expo is actually not linear like that, but sounds good!

Reply:
there's jitter/randomization involved